_posts/2024-08-06-how-to-talk-about-CHERI.markdown

```diff
@@ -0,0 +1,228 @@
+
+## So you mean compilers and ISAs can be friends?
+
+None of these language-level properties come exclusively from the hardware.
+The hardware provides tools that the compiler can choose to use.
+The compiler can also choose not to use them when it doesn't need to.
+For safe languages, a lot of properties are guaranteed by construction and so don't need enforcing *within that language's code* (or within the safe subset, for a language like Rust), though a compiler for safety-critical systems may choose to use them for defence in depth against compiler bugs.
+In C, for example, we don't enforce stack temporal safety within a compartment because it's easy for static analysers to track this kind of bug when they can see all of the code and it's a better security-performance tradeoff to recommend that people aim the gun slightly away from their foot.
+We do enforce it at compartment boundaries, because we're crossing a security boundary.
+If you choose not to run a static analyser that checks for stack temporal safety issues, you can introduce bugs into your own code, but not other compartments.
+
+The same applies for CFI properties.
+CHERIoT provides a trusted stack for cross-compartment calls and a switcher that enforces a lot of properties on both the call and return path.
+Within a compartment, we provide forward and backwards return sentries for coarse-grained CFI (you can't confuse function pointers and return addresses).
+This provides a few guarantees at library boundaries.
+You cannot jump into the middle of a library function (easily, you may be able to if it spills its program counter to the stack and doesn't zero it on return).
+You cannot call a library function that disables interrupts without using the link register to return.
+Compilers can build richer abstractions for CFI, or they can accept that memory safety plus some coarse CFI still makes life very hard for attackers.
+This choice is not made for them and, if attackers find clever techniques for starting code-reuse attacks in the future, compilers can add additional defences within a compartment.
+
+All of this is why it's important not to talk about properties that CHERI systems enforce in terms of language-level properties.
+CHERI provides tools that allow implementations to enforce language-level properties but it does not define any of these in terms of language-level constructs.
+This is particularly apparent when, for example, you create a deeply immutable pointer in C code: something that you cannot enforce with language semantics (C lets you cast away `const`) but which CHERI can enforce even on assembly code that handles that pointer.
+CHERI doesn't give C memory safety, CHERI gives you a set of tools that allow C, C++, Rust, Java, and so on to all share objects without letting any of them violate the safety properties of the others.
```

Comment:
Good to reiterate what was stated earlier using the same words: "CHERI gives you a set of tools that allow C, ... to write code that completely prevents other code from violating its safety properties."